lib/nxt_pipeline/pipeline.rb

```diff
@@ -115,22 +119,75 @@ def on_errors(*errors, halt_on_error: true, &callback)
 
     alias :on_error :on_errors
 
-    def before_execute(&callback)
-      self.before_execute_callback = callback
+    def before_step(&block)
+      callbacks.resolve!(:step, :before) << block
+    end
+
+    def after_step(&block)
+      callbacks.resolve!(:step, :after) << block
+    end
+
+    def around_step(&block)
+      callbacks.resolve!(:step, :around) << block
     end
 
-    def after_execute(&callback)
-      self.after_execute_callback = callback
+    def before_execution(&block)
+      callbacks.resolve!(:execution, :before) << block
+    end
+
+    def after_execution(&block)
+      callbacks.resolve!(:execution, :after) << block
+    end
+
+    def around_execution(&block)
+      callbacks.resolve!(:execution, :around) << block
     end
 
     private
 
+    def run_callbacks(type, kind, changeset)
+      callbacks.resolve!(type, kind).each do |callback|
+        run_callback(callback, changeset)
+      end
+    end
+
+    def run_around_callbacks(type, args, &execution)
+      around_callbacks = callbacks.resolve!(type, :around)
+      return execution.call unless around_callbacks.any?
+
+      callback_chain = around_callbacks.reverse.inject(execution) do |previous, callback|
+        -> { callback.call(self, args, previous) }
+      end
+
+      callback_chain.call
+    end
+
+    def before_execution_callbacks
```

Comment:
This method is private but I dont seem to find where it is being used? Same goes for `after_execution_callbacks`, `around_execution_callbacks` and `around_execution_callbacks`.

Reply:
Those are leftovers. I will remove.